File: otto-android/app/src/main/java/to/ottomot/driftd/routebuilder/engine/RouteAutoPathPointGenerator.kt
package to.ottomot.driftd.routebuilder.engine

/**
 * Places auto turn path markers along a snapped road polyline (ported from iOS).
 */
object RouteAutoPathPointGenerator {
    object Options {
        const val FINISH_BUFFER_METERS = 100.0
    }

    fun autoTurnPathCoordinates(
        turnCoordinates: List<RouteLatLng>,
        roadCoordinates: List<RouteLatLng>,
        polylineIndex: RoutePolylineIndex? = null,
    ): List<RouteLatLng> {
        if (roadCoordinates.size < 2) return emptyList()

        val index = polylineIndex ?: RoutePolylineIndex(roadCoordinates)
        val totalLength = RoutePolylineGeometry.polylineTotalLength(roadCoordinates)
        val projected =
            turnCoordinates.mapNotNull { turnCoordinate ->
                val projection = index.projectOntoPolyline(turnCoordinate) ?: return@mapNotNull null
                val arcLength = projection.arcLengthMeters
                if (arcLength <= Options.FINISH_BUFFER_METERS ||
                    arcLength >= totalLength - Options.FINISH_BUFFER_METERS
                ) {
                    return@mapNotNull null
                }
                projection.coordinate to arcLength
            }

        return projected
            .sortedBy { it.second }
            .map { it.first }
    }
}

File: otto-mobile/otto-mobile/RouteAutoPathPointGenerator.swift
import CoreLocation
import Foundation

enum RouteAutoPathPointGenerator {
    enum Options {
        static let finishBufferMeters = 100.0
    }

    static func autoTurnPathCoordinates(
        turnCoordinates: [CLLocationCoordinate2D],
        roadCoordinates: [CLLocationCoordinate2D],
        polylineIndex: RoutePolylineIndex? = nil
    ) -> [CLLocationCoordinate2D] {
        guard roadCoordinates.count >= 2 else { return [] }

        let index = polylineIndex ?? RoutePolylineIndex(lineCoordinates: roadCoordinates)
        let totalLength = RouteMapGeometry.polylineTotalLength(roadCoordinates)
        let projected = turnCoordinates.compactMap { turnCoordinate -> (coordinate: CLLocationCoordinate2D, arcLength: Double)? in
            guard let projection = index.projectOntoPolyline(turnCoordinate) else { return nil }
            let arcLength = projection.arcLengthMeters
            guard arcLength > Options.finishBufferMeters,
                  arcLength < totalLength - Options.finishBufferMeters else {
                return nil
            }
            return (projection.coordinate, arcLength)
        }

        return projected
            .sorted { $0.arcLength < $1.arcLength }
            .map(\.coordinate)
    }
}
